### ISSUE TITLE
zingg 0.4.0 in docker consume lots disk space

### ISSUE THREAD

**Issue by iqoOopi** [zinggAI/zingg#893]
I'm running a docker zingg with 2.7M records with 10 columns match. On my 6 core, 32gb RAM, it took 14 hours now and still running. It already has 230GB disk space. 

For the past few hours, looks like it keep writing to rdd_71_1 nonstop. 

Is this normal?

![image](https://github.com/user-attachments/assets/a0bddb32-a4b1-488e-9be6-10c61c066be9)
![image](https://github.com/user-attachments/assets/293f8585-09f9-4b8f-87c0-a30c571dcdb5)
![image](https://github.com/user-attachments/assets/3cf23aea-3ad4-4f03-8f30-6897f9552835)
![image](https://github.com/user-attachments/assets/a95f176c-3f91-45c5-8dc0-86ff82d15efd)

**Comment by sonalgoyal** [comment]
That should not happen. What is the labelDataSampleSize and number of matches you have labeled?

**Comment by iqoOopi** [comment]
labelDataSampleSize as shown is 0.001. Number of match is roughly 60 records. I trained the model with 60K records first, and runner the match with 60K. Everything works. 

Then I want to try on 2.7M records. "FindAndLabel" command works fine. But the "match" never finished and took all my disk space (more than 300GB) and never finish

**Comment by sonalgoyal** [comment]
It is better to run train on a data size close to the one you want to run match on. Can you please run train with the bigger 2.7m dataset and try match post that?

**Comment by iqoOopi** [comment]
Oh, I forgot to mention. After I switched to 2.7m, I reruned few round of "findandLabel" and retrain the model as well. Then I started the match

**Comment by sonalgoyal** [comment]
Ok. Seems it’s not trained the blocking model. Zingg jobs are more compute intensive than memory intensive, and the training samples help to learn the blocking model which helps to parallelise the job across cores. Do you have some major null values  in your dataset which may be getting clubbed together?

**Comment by iqoOopi** [comment]
Yes, each records have 15 fields. all these fields are nullable. I do have quite many records that only have value for 2 fields (firstName, phone) then rest 13 fields are null

**Comment by sonalgoyal** [comment]
Ah. That’s a tricky one. Are the null fields important to the match?

**Comment by iqoOopi** [comment]
yes, they are important if they have a value. Like (SIN number, DOB, email address etc)

**Comment by sonalgoyal** [comment]
I suspect you will need a lot more training in terms of matching data for combinations of these null values. Nulls are tough to block as they have zero signal. Maybe add through trainingSamples and see if that changes things?

**Comment by iqoOopi** [comment]
> Seems it’s not trained the blocking model.

will try train more. Btw, how do you figure out it is not using the block model?

**Comment by sonalgoyal** [comment]
I have seen it in the past on one dataset which did not have a lot of values populated. If there is no signal, it is hard to learn. I think we should build some way to let users know this while running Zingg.

**Comment by iqoOopi** [comment]
![image](https://github.com/user-attachments/assets/23540a2b-4caa-4b8f-b5ba-5df5f89352d1)
Also when running Zingg, how can I tell it is analyzing something? nothing from log. Hard to tell whether it is working normal or need abort the current task.

**Comment by sonalgoyal** [comment]
These warnings are ok. If the logs are not moving at all, that may be a sign.

**Comment by sonalgoyal** [comment]
If you are familiar with spark, you could look at the spark gui

**Comment by iqoOopi** [comment]
yeah, the log are not moving for hours. do you have a link to the spark doc where I can see the GUI? sorry new to this

**Comment by iqoOopi** [comment]
btw, in the config.json file I have "trainingSamples" and "data" section, both are pointing to SQL server table. wondering is the schema order matters?  like in training samples, I have column A,B,C, but in the data section I have column listed as A,C,B. Since they are SQL table, so I think the sequence does not matter, just want to confirm

**Comment by iqoOopi** [comment]
![image](https://github.com/user-attachments/assets/0c8d20f0-5fed-4d2b-9af4-3b99ce2fed3f)
![image](https://github.com/user-attachments/assets/e436c59c-c6d4-47c8-b05e-08254adb9c7d)

Just restart the match after re-trained for around 100 matches. 
I saw from spark GUI, it only have 1 active job with 3 tasks. Is this normal?

**Comment by sonalgoyal** [comment]
what is the numPartitions setting? For better parallelisation, you want it to be 4-5 times your number of cores

**Comment by iqoOopi** [comment]
AH, thanks @sonalgoyal  for the quick reply. In the config.json file, it is only 4.  Also in the zingg.conf file, I saw there is a "spark.default.parallelism" setting being commented out, what should be that value?

**Comment by iqoOopi** [comment]
> btw, in the config.json file I have "trainingSamples" and "data" section, both are pointing to SQL server table. wondering is the schema order matters? like in training samples, I have "schema": A string,B string,C string, but in the data section I have column "schema" as A string,C string,B string. Since they are SQL table, so I think the sequence does not matter, just want to confirm

@sonalgoyal also how about this Q?

**Comment by iqoOopi** [comment]
I make both numberOfPartitions and spark.default.parallelsim settings to 20. Is this looks normal?

![image](https://github.com/user-attachments/assets/332ca6c6-4482-4217-979f-c3a9081acf46)
![image](https://github.com/user-attachments/assets/b9d4c310-4545-4ea4-a181-d79ceb257616)
![image](https://github.com/user-attachments/assets/ab0d740b-37b3-490e-bd55-09a0d7e6c09f)

**Comment by sonalgoyal** [comment]
> > btw, in the config.json file I have "trainingSamples" and "data" section, both are pointing to SQL server table. wondering is the schema order matters? like in training samples, I have "schema": A string,B string,C string, but in the data section I have column "schema" as A string,C string,B string. Since they are SQL table, so I think the sequence does not matter, just want to confirm
> 
> @sonalgoyal also how about this Q?

I think the SQL server dataframe should be read correctly, but I am not 100% sure as thats a case we have not tested. Are you seeing an issue?

**Comment by iqoOopi** [comment]
> > > btw, in the config.json file I have "trainingSamples" and "data" section, both are pointing to SQL server table. wondering is the schema order matters? like in training samples, I have "schema": A string,B string,C string, but in the data section I have column "schema" as A string,C string,B string. Since they are SQL table, so I think the sequence does not matter, just want to confirm
> > 
> > 
> > @sonalgoyal also how about this Q?
> 
> I think the SQL server dataframe should be read correctly, but I am not 100% sure as thats a case we have not tested. Are you seeing an issue?

NO, I have not seen any issues. From the findandlabel command result, I can see the model are doing its job.

**Comment by sonalgoyal** [comment]
how did it go @iqoOopi ?

**Comment by sonalgoyal** [comment]
can you share the complete logs?

**Comment by sonalgoyal** [comment]
this is likely the case of a poorly formed blocking model. this can happen due to less traiing data, but the user has mentioned that they have labelled sufficiently. hard to say further without logs or sample data. @iqoOopi will you be open to a debug session on this?

**Comment by sonalgoyal** [comment]
the verifyBlocking phase should now give the ability to inspect the blocking tree
